File: sdk/php/src/Providers/Fmail.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;
use ChanhanzhanX\TempMail\Normalize;

/**
 * FMail 渠道实现 — https://fmail.men
 *
 * 创建 GET /v1/random（可选 ?domain=），取 address；
 * 读信 GET /v1/inbox/{local}?domain=xxx&limit=50，逐封 GET /v1/email/{token} 补全正文。
 */
final class Fmail
{
    private const CHANNEL = 'fmail';
    private const BASE_URL = 'https://fmail.men';

    /** @var array<string,string> */
    private const HEADERS = [
        'Accept' => 'application/json',
        'User-Agent' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 '
            . '(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36',
    ];

    /**
     * @return array<mixed>
     */
    private static function fetchJson(string $path): array
    {
        $resp = HttpClient::get(self::BASE_URL . $path, self::HEADERS);
        if ($resp->getStatusCode() >= 400) {
            throw new \RuntimeException('fmail: 请求失败 ' . $resp->getStatusCode());
        }
        return HttpClient::json($resp);
    }

    private static function normalizeDomain(?string $domain): ?string
    {
        $value = ltrim(trim((string) $domain), '@');
        return $value !== '' ? $value : null;
    }

    /**
     * @param array<mixed> $raw
     * @return array<mixed>
     */
    private static function flatten(array $raw, string $recipient): array
    {
        return [
            'id' => $raw['id'] ?? ($raw['uid'] ?? ($raw['token'] ?? '')),
            'from' => $raw['sender'] ?? ($raw['from'] ?? ''),
            'to' => $raw['recipient'] ?? ($raw['to'] ?? $recipient),
            'subject' => $raw['subject'] ?? '',
            'text' => $raw['body_text'] ?? ($raw['text'] ?? ($raw['snippet'] ?? '')),
            'html' => $raw['body_html'] ?? ($raw['html'] ?? ''),
            'date' => $raw['received_at'] ?? ($raw['timestamp'] ?? ($raw['date'] ?? '')),
            'is_read' => $raw['is_read'] ?? false,
            'attachments' => $raw['attachments'] ?? [],
        ];
    }

    public static function generate(?string $domain = null): EmailInfo
    {
        $selected = self::normalizeDomain($domain);
        $path = '/v1/random';
        if ($selected !== null) {
            $path .= '?domain=' . rawurlencode($selected);
        }
        $data = self::fetchJson($path);
        $email = trim((string) ($data['address'] ?? ''));
        if ($email === '') {
            $username = trim((string) ($data['username'] ?? ''));
            $dom = trim((string) ($data['domain'] ?? ''));
            if ($username !== '' && $dom !== '') {
                $email = $username . '@' . $dom;
            }
        }
        if ($email === '' || !str_contains($email, '@')) {
            throw new \RuntimeException('fmail: invalid random response');
        }
        return new EmailInfo(self::CHANNEL, $email);
    }

    /**
     * @return Email[]
     */
    public static function getEmails(string $email): array
    {
        $value = trim($email);
        $parts = explode('@', $value, 2);
        $local = $parts[0] ?? '';
        $domain = $parts[1] ?? '';
        if ($local === '' || $domain === '') {
            throw new \InvalidArgumentException('fmail: invalid email');
        }
        $data = self::fetchJson(
            '/v1/inbox/' . rawurlencode($local) . '?domain=' . rawurlencode($domain) . '&limit=50'
        );
        $rows = $data['emails'] ?? null;
        if (!is_array($rows)) {
            return [];
        }
        $out = [];
        foreach ($rows as $row) {
            if (!is_array($row)) {
                continue;
            }
            $token = trim((string) ($row['token'] ?? ($row['id'] ?? '')));
            if ($token === '') {
                $out[] = Normalize::email(self::flatten($row, $value), $value);
                continue;
            }
            try {
                $detail = self::fetchJson('/v1/email/' . rawurlencode($token));
                $nested = is_array($detail['email'] ?? null) ? $detail['email'] : $detail;
                $out[] = Normalize::email(self::flatten($nested, $value), $value);
            } catch (\Throwable) {
                $out[] = Normalize::email(self::flatten($row, $value), $value);
            }
        }
        return $out;
    }
}

File: sdk/csharp/src/Providers/Fmail.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Text.Json.Nodes;

namespace XxxXTeam.TempMail.Providers;

/// <summary>fmail 渠道（fmail.men）。GET JSON API，逐封拉取正文。</summary>
public static class Fmail
{
    private const string BaseUrl = "https://fmail.men";

    private static readonly Dictionary<string, string> Headers = new()
    {
        ["Accept"] = "application/json",
        ["User-Agent"] = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36",
    };

    private static JsonObject FetchJson(string path)
    {
        var resp = Http.Get($"{BaseUrl}{path}", Headers);
        resp.EnsureSuccess();
        return Json.Parse(resp.Body) as JsonObject ?? new JsonObject();
    }

    private static string? NormalizeDomain(string? domain)
    {
        var value = (domain ?? "").Trim().TrimStart('@');
        return value.Length > 0 ? value : null;
    }

    private static Dictionary<string, object?> FlattenMessage(JsonObject raw, string recipient)
    {
        return new Dictionary<string, object?>
        {
            ["id"] = Pick(raw, "id", "uid", "token"),
            ["from"] = Pick(raw, "sender", "from"),
            ["to"] = PickOr(raw, recipient, "recipient", "to"),
            ["subject"] = Pick(raw, "subject"),
            ["text"] = Pick(raw, "body_text", "text", "snippet"),
            ["html"] = Pick(raw, "body_html", "html"),
            ["date"] = Pick(raw, "received_at", "timestamp", "date"),
            ["is_read"] = (raw["is_read"] as JsonValue)?.GetValue<bool>() ?? false,
        };
    }

    /// <summary>创建临时邮箱（可指定域名）</summary>
    public static EmailInfo Generate(string? domain = null)
    {
        var selected = NormalizeDomain(domain);
        var path = "/v1/random";
        if (selected is not null) path = $"{path}?domain={WebUtility.UrlEncode(selected)}";
        var data = FetchJson(path);
        var email = (data["address"]?.GetValue<string>() ?? "").Trim();
        if (email.Length == 0)
        {
            var username = (data["username"]?.GetValue<string>() ?? "").Trim();
            var dom = (data["domain"]?.GetValue<string>() ?? "").Trim();
            if (username.Length > 0 && dom.Length > 0) email = $"{username}@{dom}";
        }
        if (email.Length == 0 || !email.Contains('@')) throw new Exception("fmail: invalid random response");
        return new EmailInfo("fmail", email);
    }

    /// <summary>获取邮件列表，逐封补全正文</summary>
    public static List<Email> GetEmails(string email)
    {
        var value = (email ?? "").Trim();
        var idx = value.IndexOf('@');
        if (idx <= 0 || idx == value.Length - 1) throw new Exception("fmail: invalid email");
        var local = value[..idx];
        var domain = value[(idx + 1)..];

        var data = FetchJson($"/v1/inbox/{WebUtility.UrlEncode(local)}?domain={WebUtility.UrlEncode(domain)}&limit=50");
        var result = new List<Email>();
        if (data["emails"] is not JsonArray rows) return result;

        foreach (var r in rows)
        {
            if (r is not JsonObject row) continue;
            var token = Pick(row, "token", "id").Trim();
            if (token.Length == 0)
            {
                result.Add(Normalize.NormalizeEmail(FlattenMessage(row, value), value));
                continue;
            }
            try
            {
                var detail = FetchJson($"/v1/email/{WebUtility.UrlEncode(token)}");
                var nested = detail["email"] as JsonObject ?? detail;
                result.Add(Normalize.NormalizeEmail(FlattenMessage(nested, value), value));
            }
            catch
            {
                result.Add(Normalize.NormalizeEmail(FlattenMessage(row, value), value));
            }
        }
        return result;
    }

    private static string Pick(JsonObject o, params string[] keys)
    {
        foreach (var k in keys)
        {
            if (o[k] is JsonNode n)
            {
                var s = Json.NodeToString(n);
                if (!string.IsNullOrEmpty(s)) return s;
            }
        }
        return "";
    }

    private static string PickOr(JsonObject o, string fallback, params string[] keys)
    {
        var v = Pick(o, keys);
        return v.Length > 0 ? v : fallback;
    }
}
